uni/lib/json.icn

```diff
@@ -991,3 +998,233 @@ end
+               if *val > maxLength then {
+                  return val || " is too big. Must be less than or equal to " || maxLength || " characters."
+                  }
+               }
+	    minLength := x["minLength"]
+            if \minLength then write(space, "    minLength: ", left(minLength, 20))
+	    pattern := x["pattern"]
+            if \pattern then write(space, "    pattern: ", left(pattern, 20))
+            return z
+            }
+         }
+      "integer" | "number" :{
+         if type(val) ~== "integer" & x["type"] == "integer" then {
+            return "expected integer type, got: " || val
+            }
+         else if type(val) ~== "real" & x["type"] == "number" then{
+            return "expected a number, got " || val
+            }
+         else {
+            enum := x["enum"]
+            if \enum then write(space, "   enum: ", fullimage(enum))
+            const := x["const"]
+            if \const then write(space, "    const: ", left(const, 20))
+            multipleOf := x["multipleOf"]
+            if \multipleOf then {
+               if val % mulitpleOf ~= 0 then {
+                  return val || " must be a multiple of: " || multipleOf
+                  }
+               }
+            maximum := x["maximum"]
+            if \maximum then {
+               if val > maximum then{
+                  return val || " is greater than the maximum: " || maximum
+                  }
+               }
+            exclusiveMaximum := x["exclusiveMaximum"]
+            if \exclusiveMaximum then {
+               if val >= exlusiveMaximum then{
+                  return val || " is greater than or equal to the exclusiveMaximum: " || exclusiveMaximum
+                  }
+               }
+            minimum := x["minimum"]
+            if \minimum then{
+               if val < minimum then {
+                 return val || " is less than the minimum: " || minimum
+                  }
+               }
+            exclusiveMinimum := x["exclusiveMinimum"]
+            if \exclusiveMinimum then {
+               if val <= exclusiveMinimum then {
+                  return val || " is less than the exclusiveMinimum: " || minimum
+                  }
+               }
+            }
+         return z
+         }
+
+      "&null" :{
+         return "additional properties? " || x["type"]
+         }
+      default: return z
+      }
+end
+
+#<p>
+# helper function
+# <[param obj name of the object]>
+# <[param sch name of  the schema file]>
+# <[param schRoot name of the schema file once more]>
+# The procedure _jsvalidation recusively goes into the object file and the corresponding level of the Schema file.
+# In each level the procedure  gets the value of each propertiy in the object file and calls the  validate_property() procedure.
+# with the value and the corresponding schema property.
+#
+# <[returns "" when valid, or the error string]>
+#
+#</p>
+
+procedure _jsvalidation(obj, sch, schRoot)
```

Comment:
It's probably too large a change ... but
This procedure only returns the first error it finds, leading to many cycles of "find first error" then "fix first error".

Would it be possible to suspend, rather than return and, in the calling procedures, generate a list of errors rather than just a single error?

Reply:
Yes, that makes sense. I think I will defer this for now though.